File: app/src/main/java/vinhlong/ditagis/com/khaosatdongho/async/QueryDMVatTuAsync.java
package vinhlong.ditagis.com.khaosatdongho.async;

import android.app.Activity;
import android.os.AsyncTask;
import android.support.design.widget.BottomSheetDialog;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.esri.arcgisruntime.concurrent.ListenableFuture;
import com.esri.arcgisruntime.data.Feature;
import com.esri.arcgisruntime.data.FeatureQueryResult;
import com.esri.arcgisruntime.data.QueryParameters;
import com.esri.arcgisruntime.data.ServiceFeatureTable;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.concurrent.ExecutionException;

import vinhlong.ditagis.com.khaosatdongho.R;


/**
 * Created by ThanLe on 4/16/2018.
 */

public class QueryDMVatTuAsync extends AsyncTask<Void, ArrayList<Feature>, Void> {
    public interface AsyncResponse {
        void processFinish(ArrayList<Feature> output);
    }

    public AsyncResponse delegate = null;

    private BottomSheetDialog mDialog;
    private Activity mActivity;
    private ServiceFeatureTable mServiceFeatureTable;
    public  ArrayList<Feature> features;

    public QueryDMVatTuAsync(Activity activity, ServiceFeatureTable serviceFeatureTable, AsyncResponse delegate) {
        mActivity = activity;
        mServiceFeatureTable = serviceFeatureTable;
        this.delegate = delegate;
    }

    @Override
    protected Void doInBackground(Void... voids) {
        final QueryParameters queryParameters = new QueryParameters();
        final String query = "1=1";
        queryParameters.setWhereClause(query);
        final ListenableFuture<FeatureQueryResult> featureQueryResultListenableFuture = mServiceFeatureTable.queryFeaturesAsync(queryParameters, ServiceFeatureTable.QueryFeatureFields.LOAD_ALL);
        featureQueryResultListenableFuture.addDoneListener(() -> {
            try {
                FeatureQueryResult result = featureQueryResultListenableFuture.get();
                Iterator<Feature> iterator = result.iterator();
                features = new ArrayList<>();
                while (iterator.hasNext()) {
                    Feature feature = iterator.next();
                    features.add(feature);
                }
                mDialog.dismiss();
                publishProgress(features);
            } catch (InterruptedException e) {
                e.printStackTrace();
            } catch (ExecutionException e) {
                e.printStackTrace();
            }
        });
        return null;
    }

    @Override
    protected void onPreExecute() {
        super.onPreExecute();
        mDialog = new BottomSheetDialog(this.mActivity);
        LinearLayout view = (LinearLayout) mActivity.getLayoutInflater().inflate(R.layout.layout_progress_dialog, null, false);
        ((TextView) view.findViewById(R.id.txt_progress_dialog_title)).setText("Đang tải dữ liệu...");
        mDialog.setContentView(view);
        mDialog.setCancelable(false);

        mDialog.show();
    }


    @Override
    protected void onProgressUpdate(ArrayList... values) {
        super.onProgressUpdate(values);
        delegate.processFinish(values[0]);

    }
    @Override
    protected void onPostExecute(Void result) {
        super.onPostExecute(result);

    }

}

File: app/src/main/java/vinhlong/ditagis/com/khaosatdongho/async/QueryDMVatTuAsync.kt
package vinhlong.ditagis.com.khaosatdongho.async

import android.app.Activity
import android.os.AsyncTask
import com.esri.arcgisruntime.data.Feature
import com.esri.arcgisruntime.data.QueryParameters
import com.esri.arcgisruntime.data.ServiceFeatureTable
import java.util.*
import java.util.concurrent.ExecutionException


/**
 * Created by ThanLe on 4/16/2018.
 */

class QueryDMVatTuAsync(private val mActivity: Activity, private val mServiceFeatureTable: ServiceFeatureTable, delegate: AsyncResponse) : AsyncTask<Void, ArrayList<Feature>, Void>() {

    var delegate: AsyncResponse? = null

    lateinit var features: ArrayList<Feature>

    interface AsyncResponse {
        fun processFinish(dmVatTus: ArrayList<Feature>?)
    }

    init {
        this.delegate = delegate
    }

    override fun doInBackground(vararg voids: Void): Void? {
        val queryParameters = QueryParameters()
        val query = "1=1"
        queryParameters.whereClause = query
        val featureQueryResultListenableFuture = mServiceFeatureTable.queryFeaturesAsync(queryParameters, ServiceFeatureTable.QueryFeatureFields.LOAD_ALL)
        featureQueryResultListenableFuture.addDoneListener {
            try {
                val result = featureQueryResultListenableFuture.get()
                val iterator = result.iterator()
                features = ArrayList()
                while (iterator.hasNext()) {
                    val feature = iterator.next()
                    features.add(feature)
                }
                publishProgress(features)
            } catch (e: InterruptedException) {
                e.printStackTrace()
            } catch (e: ExecutionException) {
                e.printStackTrace()
            }
        }
        return null
    }


    override fun onProgressUpdate(vararg values: ArrayList<Feature>?) {
        super.onProgressUpdate(*values)
        values[0]?.let { delegate!!.processFinish(it) }

    }


}


